File: Serene/Serene.Script/Northwind/Customer/DialogUtils.cs

using jQueryApi;
using jQueryApi.UI;
using jQueryApi.UI.Widgets;
using Serenity;
using System;
using System.Runtime.CompilerServices;

namespace Serene
{
    [Imported]
    public static class DialogUtils
    {
        public static void PendingChangesConfirmation(jQueryObject element, 
            Func<bool> hasPendingChanges)
        {
            element.Bind("dialogbeforeclose", e =>
            {
                if (!e.HasOriginalEvent() || !hasPendingChanges())
                    return;

                e.PreventDefault();

                Q.Confirm("You have pending changes. Save them?", () =>
                {
                    element.Find("div.save-and-close-button").Click();
                }, new ConfirmOptions
                {
                    OnNo = () =>
                    {
                        element.Dialog().Close();
                    }
                });
            });
        }
    }
}

File: Serene/Serene.Web/Modules/Common/Helpers/DialogUtils.ts
namespace Serene.DialogUtils {
    export function pendingChangesConfirmation(element: JQuery, hasPendingChanges: () => boolean) {
        element.bind('dialogbeforeclose', function (e) {
            if (!Serenity.WX.hasOriginalEvent(e) || !hasPendingChanges()) {
                return;
            }

            e.preventDefault();
            Q.confirm('You have pending changes. Save them?',
                () => element.find('div.save-and-close-button').click(),
                {
                    onNo: function () {
                        element.dialog().dialog('close');
                    }
                });
        });
    }
    
    export function readonlyOnChange(masterField: Serenity.Widget<any>, cascadeField: Serenity.Widget<any>, valueEnable: any): void {
        DialogUtils.conditionalReadOnly(masterField, cascadeField, valueEnable);
        masterField.change(e => { DialogUtils.conditionalReadOnly(masterField, cascadeField, valueEnable); });
    }

    export function conditionalReadOnly(masterField: Serenity.Widget<any>, cascadeField: Serenity.Widget<any>, valueEnable: any): void {
        if (Serenity.EditorUtils.getValue(masterField) == valueEnable)
            Serenity.EditorUtils.setReadOnly(cascadeField, false);
        else
            Serenity.EditorUtils.setReadOnly(cascadeField, true);
    }

}
